### DISCUSSION TITLE
Release cycle and process for dependent packages (spaceros.repos)

### DISCUSSION THREAD

**Discussion by xfiderek** [space-ros/space-ros#130] [General]
The build of packages from [spaceros.repos](https://github.com/space-ros/space-ros/blob/main/spaceros.repos) is not reproducible, as vcs entries point to branches instead of commits/tags. See https://github.com/space-ros/space-ros/issues/101 for more details. I spoke about it with @Bckempa and here are some points which are not 100% clear (at least to me). Discussion points are added as separate comments.

**Comment by xfiderek** [comment]
1. How dependent packages in spaceros.repos (e.g. process_sarif) are released with new spaceros release? Do we follow the same release cycle and create tags with the same name in dependent repos? What about repos that we don't control (e.g. ament_cobra, ikos)

**Reply by ivanperez-keera** [reply]
I would say: Yes we follow the same release cycle and the tags have the same name.

IKOS is an external repo. We should point to https://github.com/NASA-SW-VnV/ikos and to the last release tag, and the `ikos` one under the `space-ros` org on github will be removed.

For `ament_cobra`, do we have a tag that we can refer to? Otherwise, can we refer to a specific commit?

**Reply by ivanperez-keera** [reply]
Currently, the latest tag in ikos is: https://github.com/NASA-SW-VnV/ikos/releases/tag/v3.2.  The next release of IKOS is planned for March 31st, 2024.

**Reply by xfiderek** [reply]
That's the status quo of `spaceros.repos` file:
Repo | Org | Version in spaceros.repos | Has tag in origin?
-- | -- | -- | --
ament_cobra | ament | main (branch) | no
ament_ikos | ament | main (branch) | no
ament_lint | ament | spaceros (branch) | no
iceoryx | eclipse-iceoryx | v2.0.2 (tag) | yes
process_sarif | spaceros | main (branch) | no

Note that currently ikos is not pulled. I meant `ament_ikos`, my bad. We can refer to specific commits or create tags in dependent repos.
So the best will be to create tag in `process_sarif`. In case of ament dependencies, do we want to point to commits?

**Reply by Bckempa** [reply]
I agree that everything in the space-ros org should release on the same cadence. We attempted that last time but since several repos didn't change they got release branches but weren't tagged at the and looking back that was a mistake.

**Comment by xfiderek** [comment]
2. How do we want to update spaceros.repos file for every release. Do we want to manually specify all the tags / agree on some convention for all packages / script it somehow

**Reply by ivanperez-keera** [reply]
I'm not sure what is meant by this. Is the issue that, upon release, we'd presumably have to modify `spaceros.repos` to point to specific versions and right after the release we undo that change to point back to `master` / `main` branches?

**Reply by xfiderek** [reply]
Yes, exactly. I just wanted to discuss if that this is the way in which we envision that. Additionally I wanted to ask if we want to update `spaceros.repos` file manually for every release. You could imagine that if we have consistent tag names in majority of repos we could do something like this in `spaceros.repos` file:
```
  ament_cobra:
    type: git
    url: https://github.com/ament/ament_cobra.git
    version: ${CURRENT_GIT_REF}
```
We would then create a script that subsitutes ${CURRENT_GIT_REF} with current branch/tag name and pushes changes. This would limit number of manual modifications, as only final `ros2.repos` file would be modified, but maybe its not worth the hustle as we are not in control of all repos.

**Reply by Bckempa** [reply]
Scripting is cool, I have an open issue or discussion somewhere with a grab-bag of other helpful script ideas, but our process is new enough (and in flux enough) that I think we can handle manually updating tags of the 5 repos in `spacers.repos` at the moment. We should also be aware of what we are pulling in, though I know that is currently undercut by how we use rosinstall to pin to the current humble versions.

**Reply by xfiderek** [reply]
Ok, i am fully onboard with your comments.

So the release process is as follows:
1. In spaceros repo, create a new branch `release-branch`, e.g. `humble:2024.1.0-release`
2. tag dependent repos (for now `process_sarif`)
3. modify `spaceros.repos` file (replace `dev branches` with `commits` and `tags`)
4. Merge `release-branch` to `main`
5. run final tests
6. Tag spaceros repo
7. Publish image
8. Modify `spaceros.repos` file (`commits, tags` -> `dev branches`) and commit to main

Is my understanding correct?

**Comment by xfiderek** [comment]
3. In case of main branch (non-release), which versions of `spaceros.repos`  packages should be pulled? How the spaceros.repos file should look like. Do we want to point to branches?

**Reply by ivanperez-keera** [reply]
Is the issue that, upon release, we'd presumably have to modify spaceros.repos to point to specific versions and right after the release we undo that change to point back to master / main branches?

**Reply by ivanperez-keera** [reply]
(I'm not saying that's the prettiest way to do it; I'm just trying to understand the observable behavior: releases pin versions, our `master` / `main`'s `spaceros.repos` points to latest releases.)

**Reply by xfiderek** [reply]
This particular question is not about **how**, but about **what** :D[^1]. I wanted to confirm that on main spaceros branch we want to keep un-pinned branches of dependent repos, same way as it is now. This makes the most sense to me, however looking at issue https://github.com/space-ros/space-ros/issues/101 it seemed an open point:

> Should develop always build with develop and pinning only added to tagged releases, or should versions always be pinned and each subproject must update their version tag when ready?

[^1]: Question no 2 is more about how

**Reply by Bckempa** [reply]
Not a universal answer, but for this stage of this project at this scale I'd favor keeping the develop branch version pinned but I can see the arguments both ways

**Comment by xfiderek** [comment]
Closing the discussion, @ivanperez-keera will update contribution guidelines with release process and branching / tagging strategy.

**Reply by Bckempa** [reply]
Opened https://github.com/space-ros/space-ros/issues/132 to track this work
